File: ios/coachlib/coachlib/Tracking/SUDSController.m
//
//  SUDSViewController.m
//  iStressLess
//


//

#import "UIAlertView+MKBlockAdditions.h"
#import "SUDSController.h"
#import "SUDSView.h"
#import "heartbeat.h"
#import "PreExerciseSudsEvent.h"
#import "PostExerciseSudsEvent.h"
#import "SymptomRef.h"

@implementation SUDSController

- (int)suds {
	[self view];
	return sudsView.rating;
};

- (void) setSuds:(int)sud {
	[self view];
	sudsView.rating = sud;
}

- (void) recordSUDS {
    int suds = sudsView.rating;
    
    [PreExerciseSudsEvent logWithPreExerciseSudsScore:suds];
    
    [heartbeat
     logEvent:@"SUDS_READING"
     withParameters:[NSDictionary dictionaryWithObjectsAndKeys:@"suds",[NSString stringWithFormat:@"%d",suds], nil]];
    
    if (suds != -1) {
        [self setVariable:@"suds" to:[NSNumber numberWithInt:suds]];
    } else {
        [self clearVariable:@"suds"];
    }
}

- (void) recordReSUDS {
    NSNumber *sudsNum =(NSNumber*)[self getVariable:@"suds"];
    int suds = sudsNum ? [sudsNum intValue] : -1;
    int newSuds = sudsView.rating;
    
    [PostExerciseSudsEvent logWithInitialSudsScore:suds withPostExerciseSudsScore:newSuds];
    
    [heartbeat
     logEvent:@"RESUDS_READING"
     withParameters:[NSDictionary dictionaryWithObjectsAndKeys:@"suds",[NSString stringWithFormat:@"%d",suds], @"resuds",[NSString stringWithFormat:@"%d",newSuds],nil]];
    
    if (newSuds != -1) {
        [self setVariable:@"resuds" to:[NSNumber numberWithInt:newSuds]];
    } else {
        [self clearVariable:@"resuds"];
    }
}

-(void)journalIt:(Content*)source {
    SymptomRef *symptom = nil;
    NSNumber *severity = nil;

    Content *symptomContent = (Content*)[self getVariable:@"symptom"];
    if (symptomContent) symptom = [symptomContent refForSymptom];

    severity = (NSNumber*)[self getVariable:@"suds"];
    
    NSMutableDictionary *addDict = [NSMutableDictionary dictionary];
    [addDict setObject:[NSDate date] forKey:@"when"];
    if (symptom) [addDict setObject:symptom forKey:@"symptom"];
    if (severity) [addDict setObject:severity forKey:@"severity"];
    
    [self navigateToContentName:@"manage"];
    [self navigateToContentName:@"journal" withData:@{@"add":addDict}];
}

- (void) configureFromContent {
	NSArray* nibViews =  [[NSBundle mainBundle] loadNibNamed:@"SUDSView" owner:self options:nil];
	sudsView = [nibViews objectAtIndex: 0];
	[sudsView retain];
	[self addRightSideView:sudsView withMargin:CGPointMake(10, 20)];
    
	[super configureFromContent];
    
    if ([[self.content getExtraString:@"resuds"] isEqualToString:@"true"]) {
        [self registerAction:@"journalIt" withSelector:@selector(journalIt:)];
        NSNumber *sudsScore = (NSNumber*)[self getVariable:@"suds"];
        if (!sudsScore) {
            /*
            NSString *label = @"Try Another Tool";
            Content *preselectedExercise = (Content*)[self getVariable:@"preselectedExercise"];
            if (preselectedExercise) {
                if (![preselectedExercise.entity.name isEqualToString:@"ExerciseCategory"]) {
                    label = @"Try This Tool Again";
                }
            }
             */
            
            [self addButtonWithText:@"Journal This" callingBlock:^{
                [self journalIt:nil];
            }];
            /*
            [self addButtonWithText:label callingBlock:^{
                [self clearVariable:@"selectedExercise"];
                [self navigateToContentName:@"exercise"];
            }];*/

            [self addButtonWithText:@"Done" callingBlock:^{
                [self clearVariables];
                [self navigateToContentName:@"manage"];
            }];
        } else {
            sudsView.rating = [sudsScore intValue];
            [self addButtonWithText:@"Next" callingBlock:^{
                [self recordReSUDS];
                [self navigateToNext];
            }];
        }
    } else {
        [self addButtonWithText:@"Skip" callingBlock:^{
            [self navigateToNext];
        }];
        [self addButtonWithText:@"Next" callingBlock:^{
            int distress = sudsView.rating;
            if (distress == -1) {
                [UIAlertView alertViewWithTitle:@"Rate Your Distress"
                                        message:@"Please either rate your distress using the meter or tap 'Skip'."
                              cancelButtonTitle:@"Ok"];
                return;
            }
            [self recordSUDS];
            int sudsScore = sudsView.rating;
            if (sudsScore >= 9) {
                [self navigateToNextContent:[self getContentWithName:@"crisis"]];
                return;
            }
            [self navigateToNext];
        }];
    }

}

- (void)dealloc {
	[sudsView release];
    [super dealloc];
}


@end

File: android/daelib/src/gov/va/contentlib/controllers/SUDSController.java
package gov.va.contentlib.controllers;

import java.util.TreeMap;

import com.flurry.android.FlurryAgent;
import com.openmhealth.ohmage.campaigns.va.ptsd_explorer.PostExerciseSudsEvent;
import com.openmhealth.ohmage.campaigns.va.ptsd_explorer.PreExerciseSudsEvent;
import com.openmhealth.ohmage.core.EventLog;

import gov.va.contentlib.content.Content;
import gov.va.contentlib.views.SUDSMeter;
import android.app.AlertDialog;
import android.content.Context;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup.LayoutParams;
import android.webkit.WebView;
import android.widget.LinearLayout;

public class SUDSController extends ContentViewController {

	SUDSMeter meter;
	
	public SUDSController(Context ctx) {
		super(ctx);
	}
	
	public void addText(String text) {
		View tv = makeTextView(text);
		meter = new SUDSMeter(getContext());
		meter.setId(100);

		LinearLayout.LayoutParams layout = new LinearLayout.LayoutParams(LayoutParams.WRAP_CONTENT,LayoutParams.WRAP_CONTENT);
		layout.leftMargin = 20;
		
		LinearLayout leftRightClientView = new LinearLayout(getContext());
		leftRightClientView.setOrientation(LinearLayout.HORIZONTAL);
		leftRightClientView.setBackgroundColor(0);
		leftRightClientView.addView(meter,layout);
		leftRightClientView.addView(tv);
		
		clientView.addView(leftRightClientView);
		tv.setNextFocusLeftId(meter.getId());
	}

    public void journalIt(Content c) {
        Long when = System.currentTimeMillis();
        Integer sudsScore = (Integer)getVariable("suds");
        Content symptom = (Content)getVariable("symptom");

        TreeMap<String,Object> map = new TreeMap<String, Object>();
        if (when != null) map.put("when",when);
        if (sudsScore != null) map.put("severity",sudsScore);
        if (symptom != null) map.put("symptom",symptom.getID());
        navigateToContentNameWithData("journal",map);
    }
	
	public void recordSUDS() {
		Integer sudsScore = getScore();

		TreeMap<String,String> map = new TreeMap<String, String>();
		map.put("suds",""+sudsScore);
		FlurryAgent.logEvent("SUDS_READING",map);
		
		PreExerciseSudsEvent e = new PreExerciseSudsEvent();
		e.preExerciseSudsScore = sudsScore == null ? -1 : sudsScore;
		EventLog.log(e);
		
		setVariable("suds", sudsScore);
	}

	public void recordReSUDS() {
		Integer sudsScore = (Integer)getVariable("suds");
		Integer resudsScore = getScore();

		TreeMap<String,String> map = new TreeMap<String, String>();
		map.put("suds",""+sudsScore);
		map.put("resuds",""+resudsScore);
		FlurryAgent.logEvent("RESUDS_READING",map);

		PostExerciseSudsEvent e = new PostExerciseSudsEvent();
		e.initialSudsScore = sudsScore == null ? -1 : sudsScore;
		e.postExerciseSudsScore = resudsScore == null ? -1 : resudsScore;
		EventLog.log(e);

		setVariable("resuds", resudsScore);
	}

	public boolean shouldUseScroller() {
		return false;
	}

	@Override
	public void buildClientViewFromContent() {
		super.buildClientViewFromContent();

        registerAction("journalIt");

		if ("true".equals(content.getStringAttribute("resuds"))) {
			Integer sudsScore = (Integer)getVariable("suds");
			if (sudsScore == null) {
				String label = "Try Another Tool";
				Content preselectedExercise = (Content)getVariable("preselectedExercise");
				if (preselectedExercise != null) {
					if (!preselectedExercise.getType().equals("ExerciseCategory")) {
						label = "Try This Tool Again";
					}
				}

				addButton(label).setOnClickListener(new OnClickListener() {
					public void onClick(View v) {
						clearVariable("resuds");
						clearVariable("selectedExercise");
						navigateToContentName("exercise");
					}
				});
				addButton("Done").setOnClickListener(new OnClickListener() {
					public void onClick(View v) {
						clearVariable("resuds");
						navigateToContentName("manage");
					}
				});
			} else {
				meter.setScore(sudsScore);
				addButton("Next").setOnClickListener(new OnClickListener() {
					public void onClick(View v) {
						recordReSUDS();
						navigateToNext();
					}
				});
			}
		} else {
			addButton("Skip").setOnClickListener(new OnClickListener() {
				public void onClick(View v) {
					clearVariable("suds");
					navigateToNext();
				}
			});
			addButton("Next").setOnClickListener(new OnClickListener() {
				public void onClick(View v) {
					Integer distress = meter.getScore();
					if (distress == null) {
						AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
						builder.setTitle("Rate Your Distress");
						builder.setMessage("Please either rate your distress using the meter or tap 'Skip'.");
						builder.setPositiveButton("Ok", null);
						builder.show();
						return;
					}
					recordSUDS();
					Integer sudsScore = meter.getScore();
					if ((sudsScore != null) && (sudsScore.intValue() >= 9)) {
						navigateToNextContent(db.getContentForName("crisis"));
						return;
					}
					navigateToNext();
				}
			});
		}
	}

	public Integer getScore() {
		return meter.getScore();
	}
	
}
